### PR TITLE
LWG-3880: Clarify `operator+=` complexity for `{chunk,stride}_view::iterator`

### LINKED ISSUE #2995: `<ranges>`: `stride_view::iterator::operator+=` could have O(N) complexity
If the iterator and sentinel types of the underlying view type don't model `sized_sentinel_for`, the `ranges::advance` call depicted in the "Effects: Equivalent to" code:
```c++
if (n > 0) {
  missing_ = ranges::advance(current_, stride_ * n, end_);
}
```
has O(n) complexity. However, all operations on conforming random-access iterators must have O(1) complexity. This can perhaps be fixed by advancing the iterator in two steps:
```c++
if (n > 0) {
  current_ += stride_ * (n - 1);
  missing_ = ranges::advance(current_, stride_, end_);
}
```
Which has complexity `O(stride_)`. (Admittedly icky, especially when `stride_` is near `n`.)

Alternatively, we could additionally require `sized_sentinel_for<sentinel_t<Base>, iterator_t<Base>>` for `iterator` to model `random_access`, which would ensure that the problematic `advance` call is O(1).

### LINKED ISSUE #3434: LWG-3880 Clarify `operator+=` complexity for `{chunk,stride}_view::iterator`
LWG-3880 Clarify `operator+=` complexity for `{chunk,stride}_view::`*`iterator`*

This appears to be related to #2995 cited in these operators:

https://github.com/microsoft/STL/blob/9ae1b3f1faada15feff9f0a0a9fa55c9dce532f5/stl/inc/ranges#L6232-L6257

https://github.com/microsoft/STL/blob/9ae1b3f1faada15feff9f0a0a9fa55c9dce532f5/stl/inc/ranges#L7182-L7207

I haven't determined whether we can simply eliminate these comments, or if further changes are required to conform to the resolution of LWG-3880.

### PR DESCRIPTION
**Pull request by JMazurkiewicz**
Implement LWG-3880:
* Remove outdated comment,
* Close #3434,
* Fix #2995.

### CODE REVIEW

File: stl/inc/ranges [hunk 1]
```diff
@@ -6681,14 +6681,8 @@ namespace ranges {
...
 #endif // _ITERATOR_DEBUG_LEVEL != 0
-
-                    // Per to-be-filed LWG issue (See GH-2995)
-                    if constexpr (sized_sentinel_for<_Base_sentinel, _Base_iterator>) {
-                        _Missing = _RANGES advance(_Current, _Count * _Off, _End);
-                    } else {
-                        _Current += static_cast<difference_type>(_Count * (_Off - 1));
-                        _Missing = _RANGES advance(_Current, _Count, _End);
-                    }
+                    _RANGES advance(_Current, _Count * (_Off - 1));
```
Comment (strega-nil-ms) [inline stl/inc/ranges:6684]: I think this should be `_Current += _Count * (_Off - 1);`, no? What's the reasoning behind using `_RANGES advance`?

Comment (JMazurkiewicz) [reply stl/inc/ranges]: > What's the reasoning behind using `_RANGES advance`?

That's what LWG-3880 said; but I agree with you `operator+=` is better here.

**Review by strega-nil-ms — APPROVED** [review]

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/ranges [hunk 2]
```diff
@@ -6681,14 +6681,8 @@ namespace ranges {
...
-
-                    // Per to-be-filed LWG issue (See GH-2995)
-                    if constexpr (sized_sentinel_for<_Base_sentinel, _Base_iterator>) {
-                        _Missing = _RANGES advance(_Current, _Count * _Off, _End);
-                    } else {
-                        _Current += static_cast<difference_type>(_Count * (_Off - 1));
-                        _Missing = _RANGES advance(_Current, _Count, _End);
-                    }
+                    _Current += static_cast<difference_type>(_Count * (_Off - 1));
+                    _Missing = _RANGES advance(_Current, _Count, _End);
```
Comment (CaseyCarter) [inline stl/inc/ranges:6685]: This changes removes the optimization we were performing for the `sized_sentinel_case`. Is there any particular reason to do so, or should we restore the code and simply remove the "LWG issue" comment? (Comment applies to both places.)

**Review by StephanTLavavej — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for investigating this issue and ensuring that we have the most conformant, efficient code! :smile_cat: :tada: :mag:

**Pull request merged**